src/u2f.c

```diff
@@ -101,6 +106,82 @@ typedef struct __attribute__((__packed__)) {
     uint8_t challenge[U2F_NONCE_LENGTH];
 } U2F_AUTHENTICATE_SIG_STR;
 
+typedef enum {
+    U2F_STATE_IDLE,
+    U2F_STATE_AUTH,
+    U2F_STATE_REGISTER,
+} u2f_state_t;
+
+/**
+ * Holds state data for the U2F_STATE_AUTH workflow.
+ */
+typedef struct {
+    /** Part of the auth we're currently processing. */
+    enum {
+        U2F_STATE_AUTH_REQUESTED,
+        U2F_STATE_AUTH_COMPLETED,
+        U2F_STATE_AUTH_ERROR,
+        U2F_STATE_AUTH_HIJACK
+    } state;
+
+    /** If state == U2F_STATE_AUTH_ERROR, this is the error code. */
+    uint16_t error_code;
+    /** Auth request that originated this process. */
+    U2F_AUTHENTICATE_REQ auth_req;
+    /** Cmd that originated this request. */
+    uint8_t cmd;
+    /** Cid that originated this request. */
+    uint32_t cid;
+    /** apdu->p1 from the original request. */
+    uint8_t apdu_p1;
+    /**
+     * U2F usage counter, incremented by the secure chip.
+     * Valid only if state == U2F_STATE_AUTH_COMPLETED.
+     */
+    uint32_t u2f_counter;
+    /**
+     * Private key for signing this auth request.
+     * Valid only if state == U2F_STATE_AUTH_COMPLETED.
+     */
+    uint8_t privkey[U2F_EC_KEY_SIZE];
+} u2f_auth_state_t;
+
+typedef struct {
+    enum {
+        U2F_STATE_REGISTRATION_REQUESTED,
+        U2F_STATE_REGISTRATION_ERROR,
+        U2F_STATE_REGISTRATION_COMPLETED
+    } state;
+    /** If state == U2F_STATE_REGISTRATION_ERROR, this is the error code. */
+    uint16_t error_code;
+    /** Auth request that originated this process. */
+    U2F_REGISTER_REQ reg_request;
+    /** Cmd that originated this request. */
+    uint8_t cmd;
+    /** Cid that originated this request. */
+    uint32_t cid;
+    /** apdu->p1 from the original request. */
+    uint8_t apdu_p1;
+    /**
+     * U2F usage counter, incremented by the secure chip.
+     * Valid only if state == U2F_STATE_AUTH_COMPLETED.
```

Comment:
AUTH could be a typo here.

a lot of this struct is the same as in the above struct, consider refactoring the common keys into a struct which you embed in `u2f_auth_state_t` and `u2f_register_state_t`

Reply:
Old code, was refactored.